nighthawk/source/client/benchmark_client_impl.cc

```diff
@@ -0,0 +1,226 @@
+
+  request_headers_.insertMethod().value(Envoy::Http::Headers::get().MethodValues.Get);
+  request_headers_.insertPath().value(uri_->path());
+  request_headers_.insertHost().value(uri_->host_and_port());
+  request_headers_.insertScheme().value(uri_->scheme() == "https"
+                                            ? Envoy::Http::Headers::get().SchemeValues.Https
+                                            : Envoy::Http::Headers::get().SchemeValues.Http);
+}
+
+bool BenchmarkClientHttpImpl::syncResolveDns() {
+  bool dns_resolved = false;
+  auto dns_resolver = dispatcher_.createDnsResolver({});
+  Envoy::Network::ActiveDnsQuery* active_dns_query_ = dns_resolver->resolve(
+      uri_->host_without_port(), Envoy::Network::DnsLookupFamily::V4Only,
+      [this, &dns_resolved, &active_dns_query_](
+          const std::list<Envoy::Network::Address::InstanceConstSharedPtr>&& address_list) -> void {
+        active_dns_query_ = nullptr;
+        ENVOY_LOG(debug, "DNS resolution complete for {} ({} entries).", uri_->host_without_port(),
+                  address_list.size());
+        if (!address_list.empty()) {
+          target_address_ =
+              Envoy::Network::Utility::getAddressWithPort(*address_list.front(), uri_->port());
+          dns_resolved = true;
+        } else {
+          ENVOY_LOG(critical, "Could not resolve host [{}]", uri_->host_without_port());
+        }
+        dispatcher_.exit();
+      });
+  // Wait for DNS resolution to complete before proceeding.
+  dispatcher_.run(Envoy::Event::Dispatcher::RunType::Block);
+  return dns_resolved;
+}
+
+bool BenchmarkClientHttpImpl::initialize(Envoy::Runtime::Loader& runtime) {
+  if (!syncResolveDns()) {
+    return false;
+  }
+
+  envoy::api::v2::Cluster cluster_config;
+  envoy::api::v2::core::BindConfig bind_config;
+
+  cluster_config.mutable_connect_timeout()->set_seconds(timeout_.count());
+  auto thresholds = cluster_config.mutable_circuit_breakers()->add_thresholds();
+
+  thresholds->mutable_max_retries()->set_value(0);
+  thresholds->mutable_max_connections()->set_value(connection_limit_);
+  thresholds->mutable_max_pending_requests()->set_value(max_pending_requests_);
+
+  Envoy::Network::TransportSocketFactoryPtr socket_factory;
+
+  if (uri_->scheme() == "https") {
+    auto common_tls_context = cluster_config.mutable_tls_context()->mutable_common_tls_context();
+    if (use_h2_) {
+      common_tls_context->add_alpn_protocols("h2");
+    } else {
+      common_tls_context->add_alpn_protocols("http/1.1");
+    }
+    auto transport_socket = cluster_config.transport_socket();
+    if (!cluster_config.has_transport_socket()) {
+      ASSERT(cluster_config.has_tls_context());
+      transport_socket.set_name(
+          Envoy::Extensions::TransportSockets::TransportSocketNames::get().Tls);
+      Envoy::MessageUtil::jsonConvert(cluster_config.tls_context(),
+                                      *transport_socket.mutable_config());
```

Comment:
Nit: the more modern way to do this in Envoy is with typed config (`Any`), so that you don't need to round-trip with JSON.